[Go Bootcamp] Second deliverable

jvillarreal-w opened the pull request.

jvillarreal-w committed:
Additional logging line.

jvillarreal-w committed:
Get Pokemon initial info from external.

jvillarreal-w committed:
Added external Pokemon details

jvillarreal-w committed:
Saved Pokemon into CSV from external

jvillarreal-w committed:
Added unit tests.

jvillarreal-w committed:
Fixed find by id unit test.

jvillarreal-w committed:
Adjusted external limit for Pokemon entries

FerVT commented on common/constants.go:
```diff
@@ -2,4 +2,5 @@ package common
 
 const (
 	InternalDataSourcePath = "pkmn.csv"
+	TestDataSourcePath     = "pkmn_test.csv"
```
You would like to keep the test related stuff within the context of the test files. It's ok to define this as a constant, but inside the appropriate `*_test.go` file

FerVT commented on domain/model/pokemon.go:
```diff
@@ -18,3 +20,20 @@ type Stats struct {
 	BaseStatTotal  uint64 `json:"base_stat_total"`
 }
+
+func (p *Pokemon) ToStringSlice() []string {
```
Very good having this utility method being implemented by the struct

FerVT commented on interface/context/context.go:
```diff
@@ -1,4 +1,4 @@
-package controller
+package context
```
There is already a go package `context`, you would probably like to name this differently to avoid conflicts

https://pkg.go.dev/context

FerVT commented on interface/controller/pkmn_controller.go:
```diff
@@ -5,26 +5,32 @@ import (
+func (pc *pokemonController) GetPokemon(c context.Context) error {
 	var p []*model.Pokemon
 
+	external_pkmn, _ := external.NewPokemonExternal().GetExternalPokemon(p, c)
```
You sould probably keep in a variable the result of `external.NewPokemonExternal()`

FerVT replied:
If you're expecting an Error there, then remember to handle it

FerVT commented on interface/external/pkmn_external.go:
```diff
@@ -0,0 +1,163 @@
+	for i, val := range response.Results {
+		id := uint64(i + 1)
+
+		details_request, err := http.Get(fmt.Sprintf("https://pokeapi.co/api/v2/pokemon/%v/", id))
```
Keep this url as a const just like the other one

FerVT commented on interface/external/pkmn_external.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+func (pe *pokemonExternal) GetExternalPokemon(p []*model.Pokemon, c context.Context) ([]*model.Pokemon, error) {
+	request, err := http.Get(fmt.Sprintf("%v?limit=50", url))
```
I would suggest reading this article to see how you can refactor a bit to be able to mock http requests. We want to unit test this function and we don't actually want to make http requests when doing so:

https://www.thegreatcodeadventure.com/mocking-http-requests-in-golang/

FerVT commented on interface/external/pkmn_external.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+func (pe *pokemonExternal) SaveExternalPokemon(p []*model.Pokemon) {
+	csvFile, err := os.Create("pkmn.csv")
```
I can't remember, but didn't you already have a package that handles csv operations? Would be better to just keep the logic that communicates with an external api here, and another package that handles every operation you could do with csv's.

That way we can have things properly separated

Also don't hardcode the csv path, we want to keep it flexible enough to be able to test this with other files

FerVT commented on interface/external/pkmn_external.go:
```diff
@@ -0,0 +1,163 @@
+	csvWriter := csv.NewWriter(csvFile)
+	csvWriter.Write([]string{"ID", "Name", "Primary Type", "Secondary Type", "Generation", "HP", "Attack", "Defense", "Special Attack", "Special Defense", "Speed", "Base Stat Total"})
+	for _, pkmnRow := range p {
+		_ = csvWriter.Write(pkmnRow.ToStringSlice())
```
if this can return an error then handle it

FerVT commented on interface/repository/pkmn_repository_test.go:
```diff
@@ -0,0 +1,33 @@
+	"github.com/jvillarreal-w/academy-go-q42021/domain/model"
+	u "github.com/jvillarreal-w/academy-go-q42021/utils"
+)
+
```
We usually like to use the assert package to compare the expected values with the ones our functions actually returned:
https://github.com/stretchr/testify

FerVT commented on interface/repository/pkmn_repository_test.go:
```diff
@@ -0,0 +1,33 @@
+	}
+}
+
+func TestPokemonRepository_FindById(t *testing.T) {
```
For testing you would like to have multiple cases to test every single(or at least most of) possible thing that can happen in your functions, maybe if you pass an invalid value the function would return an error, if you pass a value with some different values you would receive different stuff and so on.

The way to do this is that we have like an array of test cases, where every element  defines the arguments that your function takes, and things that can change per test. and then we iterate every test and run then.

Maybe this article could help https://www.myhatchpad.com/insight/go-unit-testing-at-compass/ or if needed we could schedule a mentoring session to look at how to do testing more effectively

FerVT reviewed (changes requested):
Overall is looking good!

There are some changes that would be nice to do to have it a bit better organized, also we would like to have more unit testing coverage, it can be a bit confusing how to do it extensively but if you want we can have a mentoring session to review it.

Great job so far!

jvillarreal-w committed:
Addressing PR comments:

- Changed context module to icontext.
- Made external Pokemon instance its own variable.
- Fixed instances where the API or CSV path were hardcoded.
- Modified unit tests to work with testify assertions.
